convert2rhel/checks.py

```diff
@@ -287,66 +295,82 @@ def get_most_recent_unique_kernel_pkgs(pkgs):
+        All RHEL kmods packages starts with kernel* or kmod*
+
+    For example, consider the following list of packages::
+
+        list_of_pkgs = [
+            'kernel-core-0:4.18.0-240.10.1.el8_3.x86_64',
+            'kernel-core-0:4.19.0-240.10.1.el8_3.x86_64',
+            'kmod-debug-core-0:4.18.0-240.10.1.el8_3.x86_64',
+            'kmod-debug-core-0:4.18.0-245.10.1.el8_3.x86_64
         ]
+
+    And when this function gets called with that same list of packages,
+    we have the following output::
+
+        result = get_most_recent_unique_kernel_pkgs(pkgs=list_of_pkgs)
+        print(result)
+        # (
+        #   'kernel-core-0:4.19.0-240.10.1.el8_3.x86_64',
+        #   'kmod-debug-core-0:4.18.0-245.10.1.el8_3.x86_64'
+        # )
+
+    :param pkgs: A list of package names to be analyzed.
+    :type pkgs: list[str]
+    :return: A tuple of packages name sorted and normalized
+    :rtype: tuple[str]
     """
 
     pkgs_groups = itertools.groupby(sorted(pkgs), lambda pkg_name: pkg_name.split(":")[0])
-    return tuple(
-        max(distinct_kernel_pkgs[1], key=_repos_version_key)
-        for distinct_kernel_pkgs in pkgs_groups
-        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
-    )
+    list_of_sorted_pkgs = []
+    for distinct_kernel_pkgs in pkgs_groups:
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod")):
+            list_of_sorted_pkgs.append(max(distinct_kernel_pkgs[1], key=_repos_version_key))
+
+    return tuple(list_of_sorted_pkgs)
 
 
 def _repos_version_key(pkg_name):
+    """Identify the version key in a given package name.
+
+    Consider the following pkg_name that will be passed to this function::
+
+        pkg_name = 'kernel-core-0:4.18.0-240.10.1.el8_3.x86_64'
+
+    The output of this will be a tuple containing the package version in a tuple::
+
+        result = _repos_version_key(pkg_name=pkg_name)
+        print(result)
+        # (4, 18, 0, 240, 10, 1)
+
+    The function will ignore the package name as it is not an important information here
+    and will only care about the version that is tied to it's name.
+
+    :param pkg_name: The package to extract the version
+    :type pkg_name: str
+    :return: A tuple containing the package version.
+    :rtype: tuple[int]
+    :raises SystemExit: Raises SytemExit if it can't find the version in the pkg_name.
+    """
+    # TODO(r0x0d): This should be moved to repo.py or utils.py?
     try:
         rpm_version = KERNEL_REPO_RE.search(pkg_name).group("version")
     except AttributeError:
         logger.critical(
-            "Unexpected package:\n%s\n is a source of kernel modules.",
+            "Unexpected package: %s\n Couldn't find the version of the given package.",
             pkg_name,
         )
     else:
         return tuple(
             map(
-                _convert_to_int_or_zero,
+                convert_to_int_or_zero,
                 KERNEL_REPO_VER_SPLIT_RE.split(rpm_version),
             )
         )
```

Comment:
Thinking about the issues I raised earlier with different package versions being evaluated wrongly, I begin to think we should be using `rpm.labelCompare()` to do our sorting rather than trying to do it ourselves.  This is in part because we're not just scanning for kernel packages here, but also kmod packages (whose authors might be doing non-standard things with their naming or versioning).

The code for the two functions would look like this once ported to `rpm.labelCompare`:
``` python
import functools
import rpm

def get_most_recent_unique_kernel_pkgs(pkgs):
    pkgs_groups = itertools.groupby(sorted(pkgs), lambda pkg_name: pkg_name.split(":")[0])
    list_of_sorted_pkgs = []
    for distinct_kernel_pkgs in pkgs_groups:
        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod")):
            list_of_sorted_pkgs.append(max(distinct_kernel_pkgs[1], key=functools.cmp_to_key(_package_version_cmp)))

    return tuple(list_of_sorted_pkgs)

def _package_version_cmp(pkg_1, pkg_2):
    pkg_ver_components = []
    for pkg in pkg_1, pkg_2:
        # Remove the package name and split the rest between epoch + version and release + arch
        epoch_version, release_arch =  pkg.rsplit('-', 2)[-2:]
        # Separate the (optional) epoch from the version
        epoch_version = epoch_version.split(':', 1)
        if len(epoch_version) == 1:
            epoch = '0',
            verson = epoch_version[0]
        else:
            epoch, version = epoch_version
        # Discard the arch
        release = release_arch.rsplit('.', 1)[0]
        pkg_ver_components.append((epoch, version, release))

    return rpm.labelCompare(pkg_ver_components[0], pkg_ver_components[1])
```
The splitting could also be done via a regex.  I believe we have such a regex in pkghandler.py although we would want to audit before using it (for two reasons: (1) yum and repoquery commands output epoch in different ways depending on the command [0:foo-1.0-2.el8.x86_64 vs foo-0:1.0-2.el8.x86_64].  Sometimes an epoch of 0 is omitted and sometimes it is placed in the string, and (2) some of the code we've written cares about the arch while others do not so the rege may grab arch or it might not.

@r0x0d Do you think that switching to `rpm.labelCompare` should be in this PR or a different one?

Reply:
@abadger, I think we can include this switch in this PR without a problem. Although this is a code cleanup and not a refactor, I feel that cleaning up code sometimes leads to a better way of doing things. 

I will take a closer look at the code you provided and see if I can implement it!

Reply:
@abadger I have included your piece of code in the last commit.